chore: Sync markdown to DITA, volume 2 (#1255)
kyma-project/telemetry-manager: docs/user/04-metrics.md

- If you want to use Prometheus-based metrics, you must have instrumented your application using a library like the [Prometheus client librar](https://prometheus.io/docs/instrumenting/clientlibs/), with a port in your workload exposed serving as a Prometheus metrics endpoint.

- For the instrumentation, you typically use an SDK, namely the [Prometheus client libraries](https://prometheus.io/docs/instrumenting/clientlibs/) or the [Open Telemetry SDKs](https://opentelemetry.io/docs/instrumentation/). Both libraries provide extensions to activate language-specific auto-instrumentation like for Node.js, and an API to implement custom instrumentation.

## Architecture

In the Telemetry module, a central in-cluster Deployment of an [OTel Collector](https://opentelemetry.io/docs/collector/) acts as a gateway. The gateway exposes endpoints for the [OTLP protocol](https://opentelemetry.io/docs/specs/otlp/) for GRPC and HTTP-based communication using the dedicated `telemetry-otlp-metrics` service, to which all Kyma components and users' applications send the metrics data.

> Optionally, the Telemetry module provides a DaemonSet of an OTel Collector acting as an agent. That agent pulls metrics of a workload in the [Prometheus pull-based format](https://prometheus.io/docs/instrumenting/exposition_formats) and can provide runtime-specific metrics for the workload.

Reviewer: ```suggestion
Optionally, the Telemetry module provides a DaemonSet of an OTel Collector acting as an agent. That agent can pull metrics of a workload and the Istio sidecar in the [Prometheus pull-based format](https://prometheus.io/docs/instrumenting/exposition_formats) and can provide runtime-specific metrics for the workload.
```

![Architecture](./assets/metrics-arch.drawio.svg)

1. An application (exposing metrics in OTLP) pushes metrics to the central metric gateway service.

> 2. An application (exposing metrics in Prometheus protocol) activates the agent to scrape the metrics with an annotation-based configuration.

Reviewer: Not the application activates the input..
```suggestion
2. Activate the agent to scrape the metrics of an application (exposing metrics in Prometheus protocol) with an annotation-based configuration.
```

7. The backend can run within the cluster.
8. If authentication has been set up, the backend can also run outside the cluster.
9. You can analyze the metric data with your preferred backend system.
10. The self monitor observes the metrics flow to the backend and reports problems in the MetricPipeline status.

### Metric Gateway

In a Kyma cluster, the metric gateway is the central component to which all components can send their individual metrics. The gateway collects, enriches, and dispatches the data to the configured backend. For more information, see [Telemetry Gateways](./gateways.md).

### Metric Agent

If a MetricPipeline configures a feature in the `input` section, an additional DaemonSet is deployed acting as an agent. The agent is also based on an [OTel Collector](https://opentelemetry.io/docs/collector/) and encompasses the collection and conversion of Prometheus-based metrics. Hereby, the workload puts a `prometheus.io/scrape` annotation on the specification of the Pod or service, and the agent collects it. The agent pushes all data in OTLP to the central gateway.

### Telemetry Manager

> The MetricPipeline resource is managed by Telemetry Managr, which is responsible for managing the custom parts of the OTel Collector configuration.

Reviewer: Ts ts ts, a typo from you, I thought you are close to perfect? :)
```suggestion
The MetricPipeline resource is managed by Telemetry Manager, which is responsible for managing the custom parts of the OTel Collector configuration.
```

Author: 🫨

- Whenever the configuration changes, it validates the configuration and generates a new configuration for the gateway and agent, and for each, a ConfigMap for the configuration is generated.
- Referenced Secrets are copied into one Secret that is mounted to the gateway as well.
- Furthermore, Telemetry Manager takes care of the full lifecycle of the gateway Deployment and the agent DaemonSet itself. Only if there is a MetricPipeline defined, they are deployed.

If you don't want to use the Metrics feature, simply don't set up a MetricPipeline.

## Setting up a MetricPipeline

In the following steps, you can see how to construct and deploy a typical MetricPipeline. Learn more about the available [parameters and attributes](resources/05-metricpipeline.md).

### 1. Create a MetricPipeline

To ship metrics to a new OTLP output, create a resource of the kind `MetricPipeline` and save the file (named, for example, `metricpipeline.yaml`).

This configures the underlying OTel Collector of the gateway with a pipeline for metrics. It defines that the receiver of the pipeline is of the OTLP type and is accessible with the `telemetry-otlp-metrics` service.

> he default protocol is GRPC, but you can choose HTTP instead. Depending on the configured protocol, an `otlp` or an `otlphttp` exporter is used. Ensure that the correct port is configured as part of the endpoint. Typically, port `4317` is used for GRPC and port `4318` for HTTP.

Reviewer: ```suggestion
The default protocol is GRPC, but you can choose HTTP instead. Depending on the configured protocol, an `otlp` or an `otlphttp` exporter is used. Ensure that the correct port is configured as part of the endpoint. Typically, port `4317` is used for GRPC and port `4318` for HTTP.
```

| `prometheus.io/scheme`             | `http`, `https` | If Istio is active, `https` is supported; otherwise, only `http` is available. The default scheme is `http` unless an Istio sidecar is present, denoted by the label `security.istio.io/tlsMode=istio`, in which case `https` becomes the default. | Determines the protocol used for scraping metrics — either HTTPS with mTLS or plain HTTP. |

> [!NOTE]
> The agent can scrape endpoints even if the workload is a part of the Istio service mesh and accepts mTLS communication. However, there's a constraint: For scraping through HTTPS, Istio must configure the workload using 'STRICT' mTLS mode. Without 'STRICT' mTLS mode, you can set up scraping through HTTP by applying the annotation `prometheus.io/scheme=http`. For related troubleshooting, see [Log entry: Failed to scrape Prometheus endpoint](#log-entry-failed-to-scrape-prometheus-endpoint).

### 5. Activate Runtime Metrics

> To enable collection of runtime metrics for your containers and Pods, define a MetricPipeline that has the `runtime` section enabled as input:

Author: [PR 1241](https://github.com/kyma-project/telemetry-manager/pull/1241/files?file-filters%5B%5D=.go&file-filters%5B%5D=.md&file-filters%5B%5D=.yaml&owned-by%5B%5D=NHingerl&show-viewed-files=true#)

removed "for your Pods" but I don't understand why. Shall we keep "for your containers and Pods" or delete it completely (also in Traces)?

Reviewer: We will extend the feature and will collect more then just container metrics, also volumes and nodes. That's why I think Mostafa removed it. Maybe:
```suggestion
To enable collection of runtime metrics like for containers, define a MetricPipeline that has the `runtime` section enabled as input:
```

Author: Or we simply drop it and keep Mostafa's version?

Reviewer: The sentence misses value somehow. In order to get X, do X. What X is gets explained further down only, maybe:
```suggestion
To enable collection of metrics more specific to the container runtime, define a MetricPipeline that has the `runtime` section enabled as input:
```

        value: https://backend.example.com:4317
```

The agent configures the [kubletstatsreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/kubeletstatsreceiver) for the metric groups `pod` and `container`. With that, [system metrics](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/kubeletstatsreceiver/documentation.md) related to containers and Pods are collected.

If you want to disable the collection of the Pod or container metrics, define the `resources` section in the `runtime` input.

- The following example drops the runtime Pod metrics and only collects runtime container metrics:

  ```yaml
  apiVersion: telemetry.kyma-project.io/v1alpha1
  kind: MetricPipeline
  metadata:
    name: backend
  spec:
    input:
      runtime:
        enabled: true
        resources:

>           pod:
>             enabled: false

Reviewer: Isn't one flavor of this example enough? Maybe lets just list the "container" with "enabled" true and delete the next example, the rest people can combine?
```suggestion
          pod:
            enabled: false
          container:
            enabled: true
```

Author: let's do this in another PR

Reviewer: not sure, I would expect a call to explain why you are not interested in my feedback :P

## Operations

> A MetricPipeline creates a Deployment running OTel Collector instances in your cluster. That instances serves OTLP endpoints and ships received data to the configured backend.

Reviewer: ```suggestion
A MetricPipeline runs several OTel Collector instances in your cluster. That instances serves OTLP endpoints and ships received data to the configured backend.
```

The Telemetry module ensures that the OTel Collector instances are operational and healthy at any time, for example, with buffering and retries. However, there may be situations when the instances drop metrics, or cannot handle the metric load.

> To detect and fix such situations, check the pipeline status and check out [Troubleshooting](#troubleshooting).

Author: I aligned them now

- **Load Balancing With Istio**: To ensure availability, the metric gateway runs with multiple instances. If you want to increase the maximum throughput, use manual scaling and enter a higher number of instances.
  By design, the connections to the gateway are long-living connections (because OTLP is based on gRPC and HTTP/2). For optimal scaling of the gateway, the clients or applications must balance the connections across the available instances, which is automatically achieved if you use an Istio sidecar. If your application has no Istio sidecar, the data is always sent to one instance of the gateway.
- **Unavailability of Output**: For up to 5 minutes, a retry for data is attempted when the destination is unavailable. After that, data is dropped.
- **No Guaranteed Delivery**: The used buffers are volatile. If the gateway or agent instances crash, metric data can be lost.
- **Multiple MetricPipeline Support**: The maximum amount of MetricPipeline resources is 3.

## Troubleshooting

### No Metrics Arrive at the Backend

> **Symptom**: No metrics arrive at the backend. In the MetricPipeline status, the `TelemetryFlowHealthy` condition has status **AllDataDropped**.

Reviewer: Use bullet list

**Cause**: Incorrect backend endpoint configuration (such as using the wrong authentication credentials) or the backend is unreachable.

**Remedy**:

1. Check the `telemetry-metric-gateway` Pods for error logs by calling `kubectl logs -n kyma-system {POD_NAME}`.
2. Check if the backend is up and reachable.
3. Fix the errors.

### Not All Metrics Arrive at the Backend

> **Symptom**: The backend is reachable and the connection is properly configured, but some metrics are refused. In the MetricPipeline status, the `TelemetryFlowHealthy` condition has status **SomeDataDropped**.

Reviewer: Use bullet list
